File: Dsa-day147.py
class Solution:
    def countCoveredBuildings(self, n: int, buildings: List[List[int]]) -> int:
        from collections import defaultdict
        
        rows = defaultdict(list) 
        cols = defaultdict(list)  
        
        for x, y in buildings:
            rows[x].append(y)
            cols[y].append(x)
        
        for ys in rows.values():
            ys.sort()
        for xs in cols.values():
            xs.sort()
        
        ans = 0
        for x, y in buildings:
            row_ys = rows[x]
            has_left = row_ys[0] < y
            has_right = row_ys[-1] > y
            
            col_xs = cols[y]
            has_above = col_xs[0] < x
            has_below = col_xs[-1] > x
            
            if has_left and has_right and has_above and has_below:
                ans += 1
                
        return ans

File: Dsa-day147.java
class Solution {
    public int countCoveredBuildings(int n, int[][] buildings) {
        Map<Integer, List<Integer>> rows = new HashMap<>();
        Map<Integer, List<Integer>> cols = new HashMap<>();
        
        for (int[] b : buildings) {
            int x = b[0], y = b[1];
            rows.computeIfAbsent(x, k -> new ArrayList<>()).add(y);
            cols.computeIfAbsent(y, k -> new ArrayList<>()).add(x);
        }
        
        for (List<Integer> list : rows.values()) Collections.sort(list);
        for (List<Integer> list : cols.values()) Collections.sort(list);
        
        int ans = 0;
        for (int[] b : buildings) {
            int x = b[0], y = b[1];
            
            List<Integer> rowY = rows.get(x);
            boolean hasLeft  = rowY.get(0) < y;
            boolean hasRight = rowY.get(rowY.size() - 1) > y;
            
            List<Integer> colX = cols.get(y);
            boolean hasAbove = colX.get(0) < x;
            boolean hasBelow = colX.get(colX.size() - 1) > x;
            
            if (hasLeft && hasRight && hasAbove && hasBelow) ans++;
        }
        return ans;
    }
}